apiserver: use endpoint lease reconciler as default

aojea opened the pull request:
The apiserver owns and manages the kubernetes.default service.

It has 3 different options to reconcile the endpoints that belong to
that service:

- None: endpoints are handled by an external party.
- MasterCount: legacy, it reconciles based on the endpoints generated
and a flag specifying the number of master on the cluster.
- [Lease: default since 1.11,](https://github.com/kubernetes/kubernetes/issues/57617) each apiserver writes a lease in etcd
and renews periodically, the endpoints are generated based on the
existing leases.

It seems that when the default was set for the lease reconciler, the
controlplane code wasn't updated and kept using the master count
reconciler.

This also starts the deprecation of the master count reconciler in
favor of the lease reconciler.

The master count has different problems that are solved by the lease reconciler and
is more complex to adapt to adapt and evolve, per example, use of dual-stack addresses on the
apiserver or reconciliation of the endpoints if they were externally modified.

/kind bug
/kind cleanup
/kind deprecation

```release-note
apiserver: start deprecation of the endpoint-reconciler-type=master-count in favor of the lease reconciler.
```

aojea committed:
apiserver: use endpoint lease reconciler as default

The apiserver owns and manages the kubernetes.default service.

It has 3 different options to reconcile the endpoints that belong to
that service:

- None: endpoints are handled by an external party.
- MasterCount: legacy, it reconciles based on the endpoints generated
and a flag specifying the number of master on the cluster.
- Lease: default since 1.11, each apiserver writes a lease in etcd
and renews periodically, the endpoints are generated based on the
existing leases.

It seems that when the default was set for the lease reconciler, the
controlplane code wasn't updated and kept using the master count
reconciler.

This also starts the deprecation of the master count reconciler in
favor of the lease reconciler.

aojea commented:
/assign @liggitt @sttts @deads2k @lavalamp 
/cc @danwinship @pohly

aojea commented on pkg/controlplane/instance.go:
```diff
@@ -274,9 +274,9 @@ func (c *Config) createEndpointReconciler() reconcilers.EndpointReconciler {
+	case reconcilers.MasterCountReconcilerType:
 		return c.createMasterCountReconciler()
-	case reconcilers.LeaseEndpointReconcilerType:
+	case "", reconcilers.LeaseEndpointReconcilerType:
```
this was causing the integration framework to pick the master count reconciler by default

pohly commented:
Is this supposed to avoid the error message that I removed in https://github.com/kubernetes/kubernetes/pull/107677?

The error that I see in current master is this:
```
$ PATH=third_party/etcd:$PATH make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-alsologtostderr=false -logtostderr=false -run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling/SchedulingBasic/500Nodes"
....
+++ [0204 18:01:07] Running tests without code coverage 
E0204 18:01:14.667896  842579 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
...
```

With this PR it is different, but there's still an error:
```
E0204 17:57:57.129603  839341 controller.go:201] Unable to remove endpoints from kubernetes service: no master IPs were listed in storage, refusing to erase all endpoints for the kubernetes service
```

My expectation is that errors only get logged when something really went wrong. According to my analysis in https://github.com/kubernetes/kubernetes/commit/be4bf5cb20c18fa9b74e5aff7fc64639b1797bd3, nothing went wrong and Controller.Start just didn't interpret the nil error return from masterCountEndpointReconciler.RemoveEndpoints correctly.

aojea commented:
> Is this supposed to avoid the error message that I removed in #107677?
> 
> The error that I see in current master is this:
> 
> ```
> $ PATH=third_party/etcd:$PATH make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-alsologtostderr=false -logtostderr=false -run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling/SchedulingBasic/500Nodes"
> ....
> +++ [0204 18:01:07] Running tests without code coverage 
> E0204 18:01:14.667896  842579 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
> ...
> ```
> 
> With this PR it is different, but there's still an error:
> 
> ```
> E0204 17:57:57.129603  839341 controller.go:201] Unable to remove endpoints from kubernetes service: no master IPs were listed in storage, refusing to erase all endpoints for the kubernetes service
> ```
> 
> My expectation is that errors only get logged when something really went wrong. According to my analysis in [be4bf5c](https://github.com/kubernetes/kubernetes/commit/be4bf5cb20c18fa9b74e5aff7fc64639b1797bd3), nothing went wrong and Controller.Start just didn't interpret the nil error return from masterCountEndpointReconciler.RemoveEndpoints correctly.

I will take a look at that error later, there can be more corner cases, but it was addressing an issue when users restored etcd and the lease was stale ... I will open an issue and assign to me

danwinship commented:
> It seems that when the default was set for the lease reconciler, the
> controlplane code wasn't updated and kept using the master count
> reconciler.

Ah, to clarify, the issue is that `cmd/kube-apiserver` defaults to `lease`, but if you use `pkg/controlplane` directly (like in the integration tests), it does not.

aojea commented:
> Is this supposed to avoid the error message that I removed in #107677?
> 
> The error that I see in current master is this:
> 
> ```
> $ PATH=third_party/etcd:$PATH make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-alsologtostderr=false -logtostderr=false -run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling/SchedulingBasic/500Nodes"
> ....
> +++ [0204 18:01:07] Running tests without code coverage 
> E0204 18:01:14.667896  842579 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
> ...
> ```
> 
> With this PR it is different, but there's still an error:
> 
> ```
> E0204 17:57:57.129603  839341 controller.go:201] Unable to remove endpoints from kubernetes service: no master IPs were listed in storage, refusing to erase all endpoints for the kubernetes service
> ```
> 
> My expectation is that errors only get logged when something really went wrong. According to my analysis in [be4bf5c](https://github.com/kubernetes/kubernetes/commit/be4bf5cb20c18fa9b74e5aff7fc64639b1797bd3), nothing went wrong and Controller.Start just didn't interpret the nil error return from masterCountEndpointReconciler.RemoveEndpoints correctly.

found the problem, the integration framework does not do a clean shutdown, and when the reconcile logic tries to delete the lease etcd is down or the data was wiped

aojea commented:
it seems it has some history :) https://github.com/kubernetes/kubernetes/pull/50690

pohly commented:
> it seems it has some history :) https://github.com/kubernetes/kubernetes/pull/50690

That seems to be about cleaning up properly. Not being able to run multiple benchmarks in a single `go test -bench=.` invocation was indeed another problem that I had.

aojea commented:
> That seems to be about cleaning up properly. Not being able to run multiple benchmarks in a single `go test -bench=.` invocation was indeed another problem that I had.

hmm, there is a problem on the shutdown phase of the apiserver, the apiserver stops listening at

> 0207 15:40:02.525834 2297965 secure_serving.go:246] Stopped listening on 127.0.0.1:46197                                                                                                                                                                                                                         
"https://127.0.0.1:46197/api/v1/namespaces/default/endpoints/kubernetes": dial tcp 127.0.0.1:46197: connect: connection refused                                                                          

the reconciler tries to remove the endpoints a few ms later, of course, "connection refused" because there is nothing listening

> E0207 15:40:02.571971 2297965 controller.go:201] Unable to remove endpoints from kubernetes service: Get "https://127.0.0.1:46197/api/v1/namespaces/default/endpoints/kubernetes": dial tcp 127.0.0.1:46197: connect: connection refused

leilajal commented:
/triage accepted

aojea committed:
integration: custom etcd gracefully termination

andrewsykim commented on test/integration/framework/controlplane_utils.go:
```diff
@@ -223,6 +224,11 @@ func startAPIServerOrDie(controlPlaneConfig *controlplane.Config, incomingServer
 		controlPlaneConfig.ExtraConfig.ServiceIPRange = net.IPNet{IP: netutils.ParseIPSloppy("10.0.0.0"), Mask: net.CIDRMask(24, 32)}
 	}
+
+	if controlPlaneConfig.ExtraConfig.EndpointReconcilerType == "" {
```
q: in createEndpointReconciler you fixed the default to use the lease reconciler but what's the reason for changing the integration tests to use "None"? Do none of the existing tests rely on the default Kubernetes service?

liggitt replied:
yeah, this seems questionable... does the default of `lease` not work for integration tests?

aojea replied:
I was testing this, this commits is going away, I've found the root cause and I'm working with Abu on it
https://github.com/kubernetes/kubernetes/pull/108033

aojea replied:
the apiserver shutdown is not respecting the preshutdown hooks and racing with the reconcilers

aojea replied:
PTAL now, I've updated the PR, there were many things here :upside_down_face:

tkashem committed:
apiserver: stop http server after pre shutdown hooks

aojea committed:
run preshutdown hooks on parallel

aojea commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -432,11 +432,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
+	go func() {
+		defer close(stopHttpServerCh)
+
+		<-delayedStopOrDrainedCh
+		<-preShutdownHooksHasStoppedCh
```
@pohly this was the reason that the endpoint reconciler failed on shutdown, it tried to modify the endpoint but the http server was already down

cc: @tkashem

aojea commented on test/integration/framework/etcd.go:
```diff
@@ -128,7 +129,15 @@ func RunCustomEtcd(dataDir string, customFlags []string) (url string, stopFn fun
 	cmd.Stderr = os.Stderr
 	stop := func() {
-		cancel()
+		// try to exit etcd gracefully
+		cmd.Process.Signal(syscall.SIGTERM)
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-time.After(5 * time.Second):
+				cancel()
+			}
+		}()
```
@liggitt @pohly other source of noise in integration, the context send a SIGKILL and etcd dies automatically.
etcd support graceful shutdown with SIGTERM, if it doesn´t exist we kill it after 5 seconds

aojea replied:
this only is used if you run the test with `go test` as in `go test -timeout 100s -run ^TestLivezAndReadyz$ k8s.io/kubernetes/test/integration/controlplane -v -count 1` , the CI job uses an external etcd

aojea commented:
I may be conflating too many changes, but they are all related and were found trying to set the lease reconciler by default

aojea commented:
I need to invesitgate this well, sorry for the noise, there are things that doesn´t add up
